### ISSUE TITLE
ModelBindingHelper.TryUpdateModelAsync not binding properties

### ISSUE THREAD

**Issue by Serjster** [aspnet/Mvc#4196]
I am creating a custom model binder (**IoCModelBinder**) that resolves our ViewModel's constructor, so we don't need to have all of needed interfaces injected in the controller constructor. (I know about [FromService] but our team is very property injection averse.)

So **IoCModelBinder** is registered in **Startup** **ConfigureServices** and the it's being hit when the model needs to be resolved just fine. **IoCModelBinder** instantiates the model just fine, but looking at your code for the standard binders that come with MVC 6 I tried using **ModelBindingHelper.TryUpdateModelAsync** to bind the request values to the properties in the model, but that is not working... the properties are always null is it a bug or am I missing something... here's the **IoCModelBinder** code:

    public class IoCModelBinder : IModelBinder
    {
        IServiceProvider serviceProvider;

        public Task<ModelBindingResult> BindModelAsync(ModelBindingContext bindingContext)
        {
            if (bindingContext.ModelType.Namespace == "OUR.SOLUTION.Web.ViewModels")
            {
                if (serviceProvider == null)
                    serviceProvider = bindingContext.OperationBindingContext.HttpContext.RequestServices;

                var model = serviceProvider.GetService(bindingContext.ModelType);

                ModelBindingHelper.TryUpdateModelAsync(
                    model,
                    string.Empty,
                    bindingContext.OperationBindingContext.HttpContext,
                    bindingContext.ModelState,
                    bindingContext.OperationBindingContext.MetadataProvider,
                    new CompositeModelBinder(new List<IModelBinder> { new GenericModelBinder() }),
                    bindingContext.OperationBindingContext.ValueProvider,
                    bindingContext.OperationBindingContext.InputFormatters,
                    ObjectValidator,
                    new CompositeModelValidatorProvider(new List<IModelValidatorProvider> { bindingContext.OperationBindingContext.ValidatorProvider })
                );

                return ModelBindingResult.SuccessAsync(bindingContext.ModelType.FullName, model);
            }

            return ModelBindingResult.NoResultAsync;
        }

        private IObjectModelValidator _objectValidator;
        public IObjectModelValidator ObjectValidator
        {
            get
            {
                if (_objectValidator == null)
                    _objectValidator = serviceProvider.GetService<IObjectModelValidator>();

                return _objectValidator;
            }
            set
            {
                if (value == null)
                    throw new ArgumentNullException(nameof(value));

                _objectValidator = value;
            }
        }
    }

**Comment by Serjster** [comment]
So I looked at the ServicesModelBinder and have taken the example from that:

        public Task<ModelBindingResult> BindModelAsync(ModelBindingContext bindingContext)
        {
            if (serviceProvider == null)
                serviceProvider = bindingContext.OperationBindingContext.HttpContext.RequestServices;

            if (bindingContext == null)
                throw new ArgumentNullException(nameof(bindingContext));

            var model = serviceProvider.GetRequiredService(bindingContext.ModelType);

            bindingContext.ValidationState.Add(model, new ValidationStateEntry() { SuppressValidation = true });

            return ModelBindingResult.SuccessAsync(bindingContext.ModelType.Name, model);
        }

The problem is still the same though... I can step-through and see the values I submitted in the **bindingContext.ValueProvider** but they never get bound to my model.

**Comment by rynowak** [comment]
Oh, I see - I misunderstood. You're trying to construct the object from DI, and then have model binding update it. Sorry, my mistake :crying_cat_face: 

If you're already inside the model binding system, this shouldn't be too difficult.

I'd try something like the following:

```
        public async Task<ModelBindingResult> BindModelAsync(ModelBindingContext bindingContext)
        {
            if (serviceProvider == null)
                serviceProvider = bindingContext.OperationBindingContext.HttpContext.RequestServices;

            if (bindingContext == null)
                throw new ArgumentNullException(nameof(bindingContext));

           if (bindingContext.Model != null)
           {
                  // This binder only constructs models, avoid infinite recursion.
                  return ModelBindingResult.NoResult;
           }

            var model = serviceProvider.GetRequiredService(bindingContext.ModelType);

            // Call model binding recursively to set properties
            bindingContext.Model = model;
            var result = await bindingContext.OperationBindingContext.ModelBinder(bindingContext);

            bindingContext.ValidationState[model] = new ValidationStateEntry() { SuppressValidation = true })

            return result;
        }
```

**Comment by Serjster** [comment]
Ok... making some progress... first of all, let me make a quick correction to the above code in case someone reads this conversation later on... so line:

    var result = await bindingContext.OperationBindingContext.ModelBinder(bindingContext);

should actually be:
var result = await bindingContext.OperationBindingContext.ModelBinder.BindModelAsync(bindingContext);

Right? So the new problem now is that is now throwing the following exceptions:

    {"No service for type 'System.String' has been registered."}
    {"No service for type 'System.Net.HttpStatusCode' has been registered."}
    
This is how I register the new model binder:

            services.AddMvc().AddMvcOptions(options =>
            {
                options.ModelBinders.Add(new IoCModelBinder());
            });

Thank you for your help... I can see that we are very close now.

**Comment by rynowak** [comment]
Thanks, fixed my sample. 

You should resurrect
```
if (bindingContext.ModelType.Namespace == "iUS.Benefits.Web.ViewModels")
{
}
```

These errors `{"No service for type 'System.String' has been registered."}` come from your code trying to instantiate string with DI.

I would also register your binder first
```
services.AddMvc().AddMvcOptions(options =>
{
    options.ModelBinders.Insert(0, new IoCModelBinder());
});
```

**Comment by Serjster** [comment]
Awesome, thanks for the advice, for anyone who happens to stumble across this, here's my final result:

        public async Task<ModelBindingResult> BindModelAsync(ModelBindingContext bindingContext)
        {
            if (bindingContext == null)
                throw new ArgumentNullException(nameof(bindingContext));

            if (bindingContext.Model == null && // This binder only constructs viewmodels, avoid infinite recursion.
                bindingContext.ModelType.Namespace == "iUS.Benefits.Web.ViewModels" &&
                bindingContext.ModelType.Name.ToLower().Contains("viewmodel"))
            {
                var serviceProvider = bindingContext.OperationBindingContext.HttpContext.RequestServices;
                var model = serviceProvider.GetRequiredService(bindingContext.ModelType);

                // Call model binding recursively to set properties
                bindingContext.Model = model;
                var result = await bindingContext.OperationBindingContext.ModelBinder.BindModelAsync(bindingContext);

                bindingContext.ValidationState[model] = new ValidationStateEntry() { SuppressValidation = true };

                return result;
            }

            return await ModelBindingResult.NoResultAsync;
        }

**Comment by rynowak** [comment]
This is awesome! Glad you got it working. In RC2 there will be an easier way to do this by subclassing `MutableObjectModelBinder`. The technique used here will still work (with some API changes).

**Comment by dikoga** [comment]
hi guys, was this integrated in the current release? @Serjster did you update this snippet? thanks

**Comment by Serjster** [comment]
Not sure what you mean by "did you update this snippet?"

**Comment by dikoga** [comment]
Sorry, I did not explain why I am asking. This is not working for on this current version. The ModelType class does not have anymore the IsClass or IsInterface anymore. ModelBindingContext also doesn't have the OperationBindingContext and the interface has changed too. Thanks for replying.

Koga, Diego
----------------------------

**Comment by dougbu** [comment]
> The `ModelType` class does not have anymore the `IsClass` or `IsInterface` anymore. 

That property is of type `Type`. You'll see `IsClass` and `IsInterface` when using desktop .NET. For .NET Core, use `GetTypeInfo()` e.g. `context.ModelType.GetTypeInfo().IsClass`.

> `ModelBindingContext` also doesn't have the `OperationBindingContext` and the interface has changed too.

We remove the little-used `OperationBindingContext` class and moved its important members elsewhere between RC2 and 1.0.0. See 756cd2d and #4501 for more information.

Similarly the `IModelBinder` interface changed when we split model binding into two phases and introduced `ModelBinderFactory`. See fb81a5e and #4355 for more information.

These changes were part of a general (and rather successful) drive to improve performance of the model binding system.

**Comment by Serjster** [comment]
Thanks for the update @dougbu We actually haven't moved to RTM yet because we are waiting on some libraries we depend on ( https://docs.google.com/spreadsheets/d/14D9gV9PbugBox_GpNR6VDr_RCeOcMyBEbCCNj6edJVc/edit#gid=0 ) to go RTM first.

One thing that is a problem even in the current version of our model binder is that the order of properties in the class is no longer respected before being accessed. For example, we put the ID property at the top of the ViewModel so that its value is bound first, as the properties further on will need the ID to get their value from the database or whatever. Right now our work around is to check if the ID is 0... which is not great. So, is there anyway to have the properties be filled in the order they appear in the class again? I mean I'm sure there is, but perhaps you can recommend a way to do that without me spending who knows how many hours trying to figure this out?

**Comment by dougbu** [comment]
@Serjster not sure what you mean by "no longer respected". Model binding is by default done in the order Reflection lists properties. If that has changed (doubtful), it happened in Roslyn or the CLR.

But you can override the default using e.g. `[Display(Order = 100)]` on the `ID` property. The default `ModelMetadata.Order` is `10000` unless overridden in this way.

**Comment by Serjster** [comment]
@dougbu Thanks for replying. I created a sample project demonstrating the issue here:
https://github.com/Serjster/IOCModelBinderExample

If you put a breakpoint in the getter for SelectedCustomer and enter a value > 0 into that form field, you will notice that the first time the getter is hit, the value of ID is 0, even though it is first in the class. The second time that getter is hit will now have the value that you entered.

How do I avoid that initial hit with ID of 0?

**Comment by Serjster** [comment]
Can someone take a look at the example repo in the above post and give me some ideas please?

**Comment by Eilon** [comment]
@Serjster you'd need to update to 1.0.0 RTM at this point. If it still repros in 1.0.0 RTM then we will investigate further. Unfortunately we can't look at a repro using a release that's ~8 months old.

**Comment by fdesbrosses** [comment]
@Eilon Hi Eilon, I'm trying to do the exact same thing as @Serjster, but with the RTM. 

Basically, I wanna use my DI (autofac, but it's not relevant) to instanciate my model and inject my dependencies. 
It's working fine for the instanciation part, but in case of a HTTP Post, I wanna hydrate properties from the request body (using [FromBody] on the parameter). 

In RC1, we could just call "bindingContext.OperationBindingContext.ModelBinder.BindModelAsync(bindingContext);" as @Serjster did to go through other binders, but as @dougbu said, it has disappeared. 

I've tried to instanciate several model binders manually after using my DI (complextype, body...) but with no success. The farest I could go was with the BodyModelBinder but my model could not get deserialised. 
I've managed to read data from the request body using a formatter and input reader. The only option I can think of right now is to get data from the request body stream and go through my model properties to set their value. Is it really the only way ? Can't I use existing binders or override an existing one to handle the model construction ? 

Thanx for your reply.

**Comment by dougbu** [comment]
> as @dougbu said, it has disappeared

@fdesbrosses exactly what disappeared? It remains possible to dynamically call other model binders from within the model binding system and to use `TryUpdateModelAsync()` from outside that system.

We have a fair number of model binder examples e.g. in the MVC functional test web sites. The code in #4703 for a relatively complex scenario was also up-to-date last I looked.

**Comment by fdesbrosses** [comment]
@dougbu I meant the property OperationBindingContext on the bindingContext type. I've tried to look at the test web sites examples but could not find a sample calling dynamically (without specifying them I mean) other binders. Do you have any pointers ?

**Comment by dougbu** [comment]
The properties of `OperationBindingContext` are still available elsewhere. In some cases they should be injected from DI when the binder is instantiated by its binding provider.

For dynamic binders, gist is in #4703.

**Comment by fdesbrosses** [comment]
For what it's worth, I've found a solution to my problem, so I thought I'd share. 

For the GET side, I used @Serjster solution (overriding ComplexTypeModelBinder.Create method and use IoC to instanciate my model). 
For the POST side however, it was not enough. The BodyModelBinder was already taking care of setting values from the body through formatters. SO, I just had to configure the JSON.Net contract resolver (used through the formatter) to use IoC to instanciate my model.

And voilà !
